src/components/ucp-tabs/content-manager/buttons/callbacks/download-and-install-content.ts

```diff
@@ -0,0 +1,248 @@
+import { getStore } from '../../../../../hooks/jotai/base';
+import { download } from '../../../../../tauri/tauri-http';
+import Logger from '../../../../../util/scripts/logging';
+import { showModalOk } from '../../../../modals/modal-ok';
+import { ContentInstallationStatus } from '../../state/downloads/download-progress';
+import {
+  installModule,
+  installPlugin,
+} from '../../../../../function/extensions/installation/install-module';
+import { onFsExists, removeFile } from '../../../../../tauri/tauri-files';
+import {
+  CONTENT_TAB_LOCK,
+  contentInstallationStatusAtoms,
+} from '../../state/atoms';
+import { getHexHashOfFile } from '../../../../../util/scripts/hash';
+import { BinaryModulePackageContent } from '../../../../../function/content/store/fetch';
+
+const LOGGER = new Logger('download-button.tsx');
+
+const guessTotalSize = (currentSize: number) => {
+  const steps = [
+    1000 * 500,
+
+    1000 * 1000 * 10,
+
+    1000 * 1000 * 100,
+
+    1000 * 1000 * 500,
+
+    1000 * 1000 * 1000,
+  ];
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const step of steps) {
+    if (currentSize < step) {
+      return step;
+    }
+  }
+
+  return steps.at(-1)!;
+};
+
+export const downloadAndInstallContent = async (
+  contentElement: ContentElement,
+) => {
+  const id = `${contentElement.definition.name}@${contentElement.definition.version}`;
+
+  const setStatus = (value: ContentInstallationStatus) =>
+    getStore().set(contentInstallationStatusAtoms(id), value);
+
+  const zipSources = contentElement.contents.package.filter(
+    (pc) => pc.method === 'github-binary',
+  );
+
+  if (zipSources.length === 0) {
+    setStatus({
+      action: 'error',
+      message: `No package found for this content that supports the methods: github-binary`,
+      name: contentElement.definition.name,
+      version: contentElement.definition.version,
+    });
+    return;
+  }
+
+  const src = zipSources.at(0)!;
+
+  const gameFolder = getStore().get(GAME_FOLDER_ATOM);
+
+  const cacheDir = `${gameFolder}/ucp/.cache/`;
+  if (!(await onFsExists(cacheDir))) {
+    LOGGER.msg(`.cache directory doesn't exist, creating it!`).warn();
+    await createDir(cacheDir);
+  }
+
+  const destination = `${cacheDir}${contentElement.definition.name}-${contentElement.definition.version}.zip`;
+
+  LOGGER.msg(
+    `Downloading ${contentElement.definition.name} from ${src.url} to ${destination}`,
+  ).debug();
+
+  await download(
+    src.url,
+    destination,
+    (
+      chunkSize: number,
+      currentSize: number,
+      totalSize: number,
+      currentPercent: string,
+    ) => {
+      let percentage = parseFloat(currentPercent.replaceAll('%', ''));
+      if (Number.isNaN(percentage)) {
+        if (src.size && src.size > 0) {
+          percentage = 100 * (currentSize / src.size);
+        } else {
+          // eslint-disable-next-line no-param-reassign
+          totalSize = guessTotalSize(currentSize);
```

Comment:
Without it, the download progress bar is nonsensical. Now it is ugly, but at least less non-sensical. In practice this will never happen any more as the online store file will contain file sizes for every binary download.

Reply:
I am not sure if I would not prefer if instead the total amount of loaded KB or something would be displayed instead of the guessing code loading bare. It also simplifies the code in this file.